src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs

```diff
@@ -156,6 +159,8 @@ private async Task ProcessQueueAsync()
         ILspServices? lspServices = null;
         try
         {
+            var pendingTasks = new ConcurrentDictionary<Task, CancellationTokenSource>();
```

Comment:
Can this be made simpler, and avoid a dictionary altogether, by using a linked cancallation token for this. So we have a cancellation token for the item, and one for the queue, whcih we link (on line 185). We could further link to a new "PendingWork" token. When entering the if on line 198, we cancel the previous PendingWork token, and create a new one.

Oh actually, that wouldn't let us wait on the in-progress work, would it? Well, forget that idea, but I'm leaving the comment in case someone knows how to bridge that gap.

Reply:
Going to resolve as I don't think anyone came up with a different approach. Please reactivate if you would like to discuss further.